File: data_structures/skip_list/cpp/skip_list.cpp
#include "skip_list.h"

SkipList::SkipList(uint32_t seed)
    : head_(new Node(INT32_MIN, MAX_LEVEL)), rng_(seed) {}

SkipList::~SkipList() {
    Node* cur = head_;
    while (cur) { Node* n = cur->forward[0]; delete cur; cur = n; }
}

uint32_t SkipList::next_rand() {
    rng_ ^= rng_ << 13;
    rng_ ^= rng_ >> 17;
    rng_ ^= rng_ << 5;
    return rng_;
}

int SkipList::random_level() {
    int lvl = 0;
    while (lvl < MAX_LEVEL && (next_rand() & 1)) ++lvl;
    return lvl;
}

void SkipList::insert(int value) {
    std::vector<Node*> update(MAX_LEVEL + 1);
    Node* cur = head_;
    for (int i = level_; i >= 0; --i) {
        while (cur->forward[i] && cur->forward[i]->val < value)
            cur = cur->forward[i];
        update[i] = cur;
    }
    cur = cur->forward[0];
    if (cur && cur->val == value) return;
    int lvl = random_level();
    if (lvl > level_) {
        for (int i = level_ + 1; i <= lvl; ++i) update[i] = head_;
        level_ = lvl;
    }
    Node* n = new Node(value, lvl);
    for (int i = 0; i <= lvl; ++i) {
        n->forward[i] = update[i]->forward[i];
        update[i]->forward[i] = n;
    }
    ++size_;
}

bool SkipList::search(int value) const {
    Node* cur = head_;
    for (int i = level_; i >= 0; --i)
        while (cur->forward[i] && cur->forward[i]->val < value)
            cur = cur->forward[i];
    cur = cur->forward[0];
    return cur && cur->val == value;
}

bool SkipList::remove(int value) {
    std::vector<Node*> update(MAX_LEVEL + 1);
    Node* cur = head_;
    for (int i = level_; i >= 0; --i) {
        while (cur->forward[i] && cur->forward[i]->val < value)
            cur = cur->forward[i];
        update[i] = cur;
    }
    cur = cur->forward[0];
    if (!cur || cur->val != value) return false;
    for (int i = 0; i <= level_; ++i) {
        if (update[i]->forward[i] != cur) break;
        update[i]->forward[i] = cur->forward[i];
    }
    delete cur;
    while (level_ > 0 && !head_->forward[level_]) --level_;
    --size_;
    return true;
}

int SkipList::size() const { return size_; }
bool SkipList::is_empty() const { return size_ == 0; }

File: data_structures/skip_list/python/skip_list.py
"""スキップリスト（確率的データ構造）

操作:
  insert(v)   : 平均 O(log n)
  search(v)   : 平均 O(log n)
  remove(v)   : 平均 O(log n)
  size()      : O(1)

最大レベル 16、各レベル昇格確率 0.5。
再現性のためシードを受け取る。
"""

from typing import List, Optional
import random


_MAX_LEVEL = 16


class _Node:
    __slots__ = ("val", "forward")

    def __init__(self, val: int, level: int) -> None:
        self.val = val
        self.forward: List[Optional["_Node"]] = [None] * (level + 1)


class SkipList:
    def __init__(self, seed: int = 42) -> None:
        self._head = _Node(-(1 << 31), _MAX_LEVEL)
        self._level = 0
        self._size = 0
        self._rng = random.Random(seed)

    def _random_level(self) -> int:
        lvl = 0
        while lvl < _MAX_LEVEL and self._rng.random() < 0.5:
            lvl += 1
        return lvl

    def insert(self, value: int) -> None:
        update: List[Optional[_Node]] = [None] * (_MAX_LEVEL + 1)
        cur = self._head
        for i in range(self._level, -1, -1):
            while cur.forward[i] and cur.forward[i].val < value:
                cur = cur.forward[i]
            update[i] = cur
        cur = cur.forward[0]
        if cur and cur.val == value:
            return
        lvl = self._random_level()
        if lvl > self._level:
            for i in range(self._level + 1, lvl + 1):
                update[i] = self._head
            self._level = lvl
        node = _Node(value, lvl)
        for i in range(lvl + 1):
            node.forward[i] = update[i].forward[i]
            update[i].forward[i] = node
        self._size += 1

    def search(self, value: int) -> bool:
        cur = self._head
        for i in range(self._level, -1, -1):
            while cur.forward[i] and cur.forward[i].val < value:
                cur = cur.forward[i]
        cur = cur.forward[0]
        return cur is not None and cur.val == value

    def remove(self, value: int) -> bool:
        update: List[Optional[_Node]] = [None] * (_MAX_LEVEL + 1)
        cur = self._head
        for i in range(self._level, -1, -1):
            while cur.forward[i] and cur.forward[i].val < value:
                cur = cur.forward[i]
            update[i] = cur
        cur = cur.forward[0]
        if not cur or cur.val != value:
            return False
        for i in range(self._level + 1):
            if update[i].forward[i] != cur:
                break
            update[i].forward[i] = cur.forward[i]
        while self._level > 0 and self._head.forward[self._level] is None:
            self._level -= 1
        self._size -= 1
        return True

    def size(self) -> int:
        return self._size

    def is_empty(self) -> bool:
        return self._size == 0
